src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -861,5 +862,131 @@ unsafe void M()
+                // (6,7): error CS0306: The type 'delegate*<void>' may not be used as a type argument
+                // class R2 : R1<delegate*<void>>
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "R2").WithArguments("delegate*<void>").WithLocation(6, 7)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeAsThisOfExtensionMethod()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+unsafe static class C
+{
+    static void M1(this delegate*<void> ptr) {}
+    static void M2(delegate*<void> ptr)
+    {
+        ptr.M1();
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (4,25): error CS1103: The first parameter of an extension method cannot be of type 'delegate*<void>'
+                //     static void M1(this delegate*<void> ptr) {}
+                Diagnostic(ErrorCode.ERR_BadTypeforThis, "delegate*<void>").WithArguments("delegate*<void>").WithLocation(4, 25)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeAsThisOfExtensionMethod_DefinedInIl()
+        {
+            const string ilSource = @"
+.class public auto ansi abstract sealed beforefieldinit CHelper
+    extends [mscorlib]System.Object
+{
+    .custom instance void [mscorlib]System.Runtime.CompilerServices.ExtensionAttribute::.ctor() = (
+        01 00 00 00
+    )
+    // Methods
+    .method public hidebysig static 
+        void M (
+            method void*() i
+        ) cil managed 
+    {
+        .custom instance void [mscorlib]System.Runtime.CompilerServices.ExtensionAttribute::.ctor() = (
+            01 00 00 00
+        )
+        // Method begins at RVA 0x205c
+        // Code size 9 (0x9)
+        .maxstack 8
+
+        IL_0001: ldc.i4.1
+        IL_0002: call void [mscorlib]System.Console::WriteLine(int32)
+        IL_0008: ret
+    } // end of method CHelper::M
+
+} // end of class CHelper
+";
+            const string source = @"
+static class C
+{
+    static unsafe void Main()
+    {
+        delegate*<void> ptr = null;
+        ptr.M();
+    }
+}";
+            var comp = CreateCompilationWithIL(source, ilSource, options: TestOptions.UnsafeReleaseExe, parseOptions: TestOptions.RegularPreview);
+
+            var verifier = CompileAndVerify(comp, expectedOutput: "1", verify: Verification.Skipped);
+            verifier.VerifyIL("C.Main", expectedIL: @"
+{
+  // Code size       10 (0xa)
+  .maxstack  1
+  .locals init (delegate*<void> V_0) //ptr
+  IL_0000:  ldc.i4.0
+  IL_0001:  conv.u
+  IL_0002:  stloc.0
+  IL_0003:  ldloc.0
+  IL_0004:  call       ""void CHelper.M(delegate*<void>)""
+  IL_0009:  ret
+}
+");
+        }
+
+        [Fact]
+        public void FunctionPointerTypeInAnonymousType()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+unsafe static class C
+{
+    static void M(delegate*<void> ptr)
+    {
+        var a = new { Ptr = ptr };
+        var b = new { Ptrs = new[] { ptr } };
```

Comment:
>var b = new { Ptrs = new[] { ptr } }; [](start = 8, length = 37)

I am confused. Why are we not getting an error for this line, Given your response: "We report an error for arrays of ```int*``` today" And the current implementation of ```IsUnsafe``` I expected the same error on this line.

Reply:
Because I apparently didn't save the file before committing, it's sitting here on my computer.

---
In reply to: [408285627](https://github.com/dotnet/roslyn/pull/43248#discussion_r408285627) [](ancestors = 408285627)